### PR TITLE
fix(migration): Make pfile sha calculation separate script

### PR DESCRIPTION
**Pull request by GMishx**
## Description

Since calculation of SHA256 for all pfiles is a computation heavy job and requires extra time to run, it is advisable to make it optional during initialization and allow user to run it as a separate job in system.

This PR aims to implement the same.

Since the pfile computation is optional, new flags (`--force-decision`, `--force-pfile`) are added to `fo-postinstall` and `fossinit.php`. Along with the flags, user can set `FOSSPFILE` environment variable to continue the computation of `pfile` checksums (similar to `--force-pfile`).
If non of the options are provided and the total time required by the script is less than 45 minutes, the migration will continue.
Otherwise, a message will be printed on the screen showing the options which can be used to do the migration (forcing using flags, environment variable or running different script).

### Changes

1. Make the script to calculate SHA256 of pfile as a separate script which can run without fo-postinstall.
2. Make PostgreSQL transactions more atomic.
3. Optimize update query.
4. Revert `pfile_md5_sha1_size_ukey` constraint due to unexpected behavior.

The following message will be printed on the screen if scripts requires more than 45 minutes to run:
image

## How to test

1. Install commit >= 4d03f6d6b4294af206bf63dff98b77e892d5a1f9 to remove `pfile_sha256` column from `pfile` table.
1. Run `fo-postinstall` and see if the script continues if total time required is less than 45 minutes.
1. Run `fo-postinstall` with `--force-pfile` and see if scripts continues completely.
1. Run `fo-postinstall` with `FOSSPFILE` env set.
1. Run the script returned by `fo-postinstall` (`/path/to/fossology/dbmigrate_pfile_calculate_sha256.php`) and see if it works.

### CODE REVIEW

**Comment by mcjaeger** [comment]
(from the discussion) there was a discussion about:
* user friendly
* ci run time friendly
* existing installation friendly

ideas:
* normal case is that sha256 will be performed at install time
* if projection of time will result in more than 45mins (because gitlab has 1h timeout): Error, installation script will take too much time, executing sha256 calculation is cancelled and needs to be applied after installation

**Comment by maxhbr** [comment]
How does this handle pfile entries, which correspond to files missing in the repository?

**Comment by GMishx** [comment]
> 
> 
> How does this handle pfile entries, which correspond to files missing in the repository?

Nice question, I forgot to mention it. Right now they are just marked as NULL entries in DB and need to be handled separately in the code wherever required.

File: install/db/dbmigrate_pfile_calculate_sha256.php.in [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+function explainPfileUsage()
+{
+  global $argv;
+
+  $usage = "Usage: " . basename($argv[0]) . " [options]
+  Calculate the SHA256 hash of files pointed by pfile table. Options are:
+  -c  path to fossology configuration files
+  -d  {database name} default is 'fossology'
+  -h  this help usage
```
[the remark this answers is not in the document]
Comment (GMishx) [reply install/db/dbmigrate_pfile_calculate_sha256.php.in]: Fixed. Please check.

File: install/fossinit.php [hunk 2]
```diff
@@ -53,16 +55,22 @@ function explainUsage()
...
 $DatabaseName = "fossology";
 $UpdateLiceneseRef = false;
 $sysconfdir = '';
 $delDbPattern = 'the option -rfosstest will drop data bases with datname like "fosstest%"';
+$forceDecision = false;
+$forcePfile = false;
 
 /* command-line options */
-$Options = getopt($AllPossibleOpts);
+$Options = getopt($AllPossibleOpts, $longOpts);
```
[the remark this answers is not in the document]
Comment (GMishx) [reply install/fossinit.php]: Fixed. Please check.

**Comment by shaheemazmalmmd** [comment]
Code looks good, Tested works as expected.

**Pull request merged**